README.markdown
# FoodInfo

FoodInfo is a ruby gem that retrieves food nutrition information from various online data sources.



## Installation

FoodInfo is available as a gem, so installation is as simple as:

    gem install food_info



## Supported Data Sources

There's currently only one adapter implemented, which pulls data from [FatSecret's REST API](http://platform.fatsecret.com/api/Default.aspx?screen=rapih).  The code's modular and adding additional data sources should be fairly straightforward, but since DailyBurn discontinued their API access I don't know of any other solid sources (if you do, though, please let me know and/or add an adapter!).



## Usage

### Housekeeping

To use the FatSecret API (currently your only option), you'll first need to [sign up for a free developer account](http://platform.fatsecret.com/api/Default.aspx?screen=r) and retrieve the "REST API Consumer Key" and "REST API Consumer Secret" from your "My Account" tab.

Once that's done, the first step is to tell FoodInfo which adapter you want to use and what authorization to send.

    FoodInfo.establish_connection(:fat_secret, :key => 'YOUR-KEY', :secret => 'YOUR-KEY')


### Caching

To cache results, FoodInfo supports passing in an instance of a memcache-API-compatible (i.e. responds to <code>get</code> and <code>set</code>) caching object.  I recommend using the [Dalli gem](https://github.com/mperham/dalli).

    require 'dalli'
    client = Dalli::Client.new('localhost:11211')
    FoodInfo.establish_connection(:fat_secret, :key => 'YOUR-KEY', :secret => 'YOUR-KEY', :cache => client)

With that in place repeated <code>search</code> or <code>details</code> requests will pull from the cache, and not the API endpoint.

### Searching

Now we can search for foods.

    cheese = FoodInfo.search('cheese')
    cheese.total_results    # => 2469
    cheese.per_page         # => 20
    cheese.page             # => 1
    cheese.results          # => ... big array ...
    cheese.results.first    # => 
    # {
    #     "description" => "Per 100g - Calories: 403kcal | Fat: 33.14g | Carbs: 1.28g | Protein: 24.90g",
    #              "id" => "33689",
    #            "kind" => "Generic",
    #            "name" => "Cheddar Cheese",
    #             "url" => "http://www.fatsecret.com/calories-nutrition/usda/cheddar-cheese"
    # }

(As an aside, I get that pretty, nicely-lined-up console formatting from the remarkably awesome [AwesomePrint Gem](https://github.com/michaeldv/awesome_print)).

#### Pagination

Search supports pagination via the <tt>page</tt> and <tt>per_page</tt> (max 50) parameters:

    FoodInfo.search('cheese', :page => 2, :per_page => 50)

#### Access to search results

You can access the results of the search explicitly:

    FoodInfo.search('cheese').results # => ... lots of results ...

But the SearchResults class includes Enumerable, so you can also just call enumerable methods on it directly:

    FoodInfo.search('cheese').map(&:name) # => array of names of matching foods
    

### Nutritional Details

Once you have a specific food item in mind from the search results, you can retrieve a whole lot of additional information.

    cheddar = FoodInfo.search('cheese').first
    info = FoodInfo.details( cheddar.id ) # => ... a whole lotta data ...

General metadata about the cheese includes id, name, kind, and url, which are identical to what you'd get from the <tt>search</tt> method.  It also has one or more servings, however, and this is where we finally get our nutrition info.

    serving = info.servings.first # =>
    # {
    #                     "calcium" => 95,
    #                    "calories" => 532.0,
    #                "carbohydrate" => 1.69,
    #                 "cholesterol" => 139.0,
    #                         "fat" => 43.74,
    #                       "fiber" => 0.0,
    #                          "id" => "29131",
    #                        "iron" => 5,
    #     "measurement_description" => "cup, diced",
    #       "metric_serving_amount" => 132.0,
    #         "metric_serving_unit" => "g",
    #         "monounsaturated_fat" => 12.396,
    #             "number_of_units" => 1.0,
    #         "polyunsaturated_fat" => 1.243,
    #                   "potassium" => 129.0,
    #                     "protein" => 32.87,
    #               "saturated_fat" => 27.841,
    #         "serving_description" => "1 cup diced",
    #                      "sodium" => 820.0,
    #                       "sugar" => 0.69,
    #                   "trans_fat" => 0.0,
    #                         "url" => "http://www.fatsecret.com/calories-nutrition/usda/cheddar-cheese?portionid=29131&portionamount=1.000",
    #                   "vitamin_a" => 26,
    #                   "vitamin_c" => 0
    # }
    
For full details on what each of those fields contains, check [the FatSecret documentation](http://platform.fatsecret.com/api/Default.aspx?screen=rapiref&method=food.get#methodResponse).


## Legal Note

The FatSecret TOS requires you not to store, well, [pretty much anything](http://platform.fatsecret.com/api/Default.aspx?screen=rapisd) aside from food or serving IDs for more than 24 hours.  This is annoying, but I figured I'd give you a heads up.


## Note on Patches/Pull Requests

Contributions are welcome, particularly adding adapters for additional data sources.

As always, the process is to fork this project on Github, make your changes (preferably in a topic branch, and without changing the gem version), send a pull request, and then receive much appreciation!

## License

Copyright &copy; 2011 [Deviantech, Inc.](http://www.deviantech.com) and released under the MIT license.



HISTORY.markdown
# Changelog

## 0.0.7 December 5, 2013
* Pulling in misc bugfixes from eclipxe
* Adding caching: can pass in memcache-compatible object to use caching

## 0.0.6 October 21, 2011
* FatSecret: Retry once if timestamp invalid (API only supports single request/second)

## 0.0.5 September 26, 2011
* Bugfix: missing 'return' was wreaking havoc with FoodInfo.search

## 0.0.4 September 26, 2011
* API improvement: Can index directly into search results (e.g. FoodInfo.search('a')[3])
* Bugfix: FoodInfo.details('..') now correctly handles results with only one serving returned

## 0.0.3 September 12, 2011
* Added missing gem dependency on ruby-hmac

## 0.0.2 September 11, 2011
* Added propagation of errors from data source
* API improvement: No need to call <tt>results</tt> on FoodItem.search to access search results

## 0.0.1  September 10, 2011
* Initial release. FatSecret adapter supports both <tt>search</tt> and <tt>details</tt> methods.

lib/food_info.rb
require 'hashie'

require "food_info/utils"
require "food_info/errors"
require "food_info/adapters"
require "food_info/cache_adapters"
require "food_info/version"


module FoodInfo
  
  # Allow extending to additional data sources in the future
  # Each adapter should implement +search+ and +details+ methods.
  ADAPTERS = {:fat_secret => FoodInfo::Adapters::FatSecret}
  
  class << self

    # Sets the adapter we'll be pulling data from.
    #
    # Example usage:
    #   FoodInfo.establish_connection(:fat_secret, :key => '...', :secret => '...')
    #
    # Example with cache:
    #   require 'dalli' # Use dalli gem to interface with memcache
    #   client = Dalli::Client.new('localhost:11211')
    #   FoodInfo.establish_connection(:fat_secret, :key => '...', :secret => '...', :cache => client)
    def establish_connection(adapter_name, opts = {})
      klass = ADAPTERS[adapter_name.to_sym]
      raise UnsupportedAdapter.new("Requested adapter ('#{adapter_name}') is unknown") unless klass
      
      # Set up the pool of workers (net yet implemented, so defaulting to "pool" of one)
      @@pool = []
      @@cursor = 0
      (opts.delete(:pool) || 1).to_i.times do
        @@pool << klass.new(opts) 
      end
      
      # Set up the cache, if any provided
      obj = opts.delete(:cache)
      @cache = obj ? FoodInfo::CacheAdapters::MemCacheCompatible.new(obj) : FoodInfo::CacheAdapters::Default.new
      
      true
    end

    # ======================
    # = Public API Methods =
    # ======================
    def search(q, opts = {})
      cached(:search, q, opts)
    end
    
    def details(id, opts = {})
      cached(:details, id)
    end


    # ==========================
    # = Implementation details =
    # ==========================
    def cached(method, param, opts = {})
      key = request_key(method, param, opts)
      
      @cache.get(key) || @cache.set(key, next_adapter.send(method, param, opts))
    end

    # FUTURE: This connection pool code won't do much good until HTTParty is non-blocking
    def next_adapter
      raise NoAdapterSpecified.new("You must run FoodInfo.establish_connection first") unless defined?(@@pool)
      @@cursor = (@@cursor + 1) % @@pool.length
      @@pool[@@cursor]
    end
    
    # Convert method + args into a string for use as the cache key
    def request_key(method, param, opts = {})
      # {:param => 123, :other => "something"} # => "other=something:param=123"
      str_opts = opts.sort{|a,b| a.to_s <=> b.to_s}.map{|pair| pair.join('=')}.join(':')
      
      "FoodInfo:#{method}:#{param}:#{str_opts}"
    end
    
  end
end


__END__

FoodInfo.establish_connection(:fat_secret, :key => ENV['KEY'], :secret => ENV['SECRET'])
a=FoodInfo.search('cheese', :page => 1, :per_page => 1).results.first
a=FoodInfo.search('cheese')

FoodInfo.establish_connection(:fat_secret, :key => ENV['KEY'], :secret => ENV['SECRET'])
a=FoodInfo.details("33689")
